File: backend/src/modules/gateway/protocols/_shared/error-payload.ts
export function parseJsonObjectErrorPayload(text: string, headers: Headers): {
  payload: Record<string, unknown>
  error: Record<string, unknown>
} | undefined {
  const trimmed = text.trim()
  if (!headers.get('content-type')?.includes('json') && !trimmed.startsWith('{')) return undefined
  try {
    const payload = JSON.parse(trimmed) as unknown
    if (!isRecord(payload)) return undefined
    const error = isRecord(payload.error) ? payload.error : payload
    return { payload, error }
  } catch {
    return undefined
  }
}

export function stringErrorField(value: unknown): string | undefined {
  if (typeof value === 'string') {
    const text = value.trim()
    return text || undefined
  }
  if (typeof value === 'number' || typeof value === 'boolean') {
    return String(value)
  }
  return undefined
}

export function firstErrorFieldText(...values: unknown[]): string | undefined {
  for (const value of values) {
    const text = errorFieldText(value)
    if (text) return text
  }
  return undefined
}

export function nestedErrorObject(value: unknown): Record<string, unknown> | undefined {
  if (!isRecord(value)) return undefined
  for (const key of ['error', 'err', 'detail', 'details', 'data']) {
    const child = value[key]
    if (isRecord(child)) return child
  }
  return undefined
}

function errorFieldText(value: unknown): string | undefined {
  const scalar = stringErrorField(value)
  if (scalar) return scalar
  if (!isRecord(value)) return undefined
  return firstErrorFieldText(
    value.message,
    value.msg,
    value.error_message,
    value.error_description,
    value.detail,
    value.reason,
    value.code,
    value.type
  )
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

File: backend-go/projects/gateway/internal/gatewayanthropic/errorpayload.go
package gatewayanthropic

import (
	"encoding/json"
	"net/http"
	"strings"
)

// ErrorPayload 对齐 GatewayProtocolErrorPayload（code/type/message 三字段契约）。
type ErrorPayload struct {
	Code    string
	Type    string
	Message string
}

// IsZero 表示上游响应没有可解析的错误负载。
func (p ErrorPayload) IsZero() bool { return p.Code == "" && p.Type == "" && p.Message == "" }

// ParseErrorPayload 对齐 parseAnthropicErrorPayload：仅解析 JSON 负载
// （Content-Type 含 json 或文本以 { 开头）。
func ParseErrorPayload(text string, header http.Header) ErrorPayload {
	parsed, ok := parseJSONObjectErrorPayload(text, header)
	if !ok {
		return ErrorPayload{}
	}
	return anthropicErrorPayloadFromParsed(parsed.payload, parsed.errorObject)
}

// ParseErrorPayloadFromJSONValue 对齐 parseAnthropicErrorPayloadFromJsonValue。
func ParseErrorPayloadFromJSONValue(value any) ErrorPayload {
	payload, errorObject, ok := jsonObjectErrorPayload(value)
	if !ok {
		return ErrorPayload{}
	}
	return anthropicErrorPayloadFromParsed(payload, errorObject)
}

func anthropicErrorPayloadFromParsed(payload, errorObject map[string]any) ErrorPayload {
	if payload == nil {
		return ErrorPayload{}
	}
	typeValue := orString(stringFieldValue(errorObject["type"]), stringFieldValue(payload["type"]))
	return ErrorPayload{
		Code: orString(
			stringFieldValue(errorObject["code"]),
			stringFieldValue(payload["code"]),
			typeValue,
		),
		Type: typeValue,
		Message: orString(
			stringFieldValue(errorObject["message"]),
			stringFieldValue(errorObject["msg"]),
			stringFieldValue(errorObject["error_message"]),
			stringFieldValue(errorObject["error_description"]),
			stringFieldValue(errorObject["detail"]),
			stringFieldValue(payload["message"]),
			stringFieldValue(payload["msg"]),
			stringFieldValue(payload["error_message"]),
			stringFieldValue(payload["error_description"]),
			stringFieldValue(payload["detail"]),
		),
	}
}

// parsedErrorPayload 持有根对象与其中的 error 子对象（error 缺省时为根对象）。
type parsedErrorPayload struct {
	payload     map[string]any
	errorObject map[string]any
}

// parseJSONObjectErrorPayload 对齐 _shared/error-payload.ts 的
// parseJsonObjectErrorPayload。
func parseJSONObjectErrorPayload(text string, header http.Header) (parsedErrorPayload, bool) {
	trimmed := strings.TrimSpace(text)
	contentType := ""
	if header != nil {
		contentType = header.Get("Content-Type")
	}
	if !strings.Contains(contentType, "json") && !strings.HasPrefix(trimmed, "{") {
		return parsedErrorPayload{}, false
	}
	var value any
	if err := json.Unmarshal([]byte(trimmed), &value); err != nil {
		return parsedErrorPayload{}, false
	}
	payload, errorObject, ok := jsonObjectErrorPayload(value)
	if !ok {
		return parsedErrorPayload{}, false
	}
	return parsedErrorPayload{payload: payload, errorObject: errorObject}, true
}

// jsonObjectErrorPayload 对齐 jsonObjectErrorPayload。
func jsonObjectErrorPayload(value any) (map[string]any, map[string]any, bool) {
	payload, ok := value.(map[string]any)
	if !ok {
		return nil, nil, false
	}
	errorObject, ok := payload["error"].(map[string]any)
	if !ok {
		errorObject = payload
	}
	return payload, errorObject, true
}
